ModBiome! (and some reorganization of various fx)

Agrair opened the pull request:
### What is the new feature?
`ModBiome`. Specifically, all biome related hooks from `ModPlayer` have been moved into a new type.
Not only that, but quite a few other hooks relating to "atmospheric" functionality have been centralized and shuffled around. Please see solxan's more in depth comment below on this thread.

### Why should this be part of tModLoader?
It further encourages the reduction of godclassing first of all, but now there are no more ugly `ZoneMyNewFancyBiome` `bool`s (at least, they were ugly in my opinion, and I hope others agree with me). There is now just `player.Active(ModContent.GetInstance<MyNewFancyBiome>()`, and tML will handle the rest.
As for the atmospheric functionality changes, a lot of priority management has been coalesced in to what will hopefully be a more manageable state.

### Are there alternative designs?
Open to suggestions.

### Sample usage for the new feature
N/A (Not going to put anything here because I hope it's really obvious what the usage is)

### ExampleMod updates
None, yet. If enough people approve this PR , then someone could port the old Example Biome stuff (I think kittenchilly's mod generation PR may interfere with this as he also did some bits with Example Biome)

Commit:
moved all biome related hooks from ModPlayer to ModBiome

Agrair commented on patches/tModLoader/Terraria/ModLoader/ModBiome.cs:
```diff
@@ -0,0 +1,68 @@
+			BiomeLoader.Add(this);
+		}
+
+		internal void Update() {
```
This concept of `Active` and the internalized `Update` method differ from I think all current hook types if I'm not mistaken, but it was the best I could come up with. I'm open to suggestions for changes.

Agrair commented on patches/tModLoader/Terraria/ModLoader/BiomeLoader.cs:
```diff
@@ -0,0 +1,142 @@
+				writer.Write(data);
+				return true;
+			}
+			return false;
```
I just realized, this is probably creating a little bit of overhead. This could be replaced with
```
if (modBiome.Active)
{
    writer.Write(modBiome.Mod.Name);
    writer.Write(modBiome.Name);
    writer.Write(sizeof(bool));
    writer.Write(modBiome.Active);
}
```
Maybe not even that. I can't remember exactly, but do `false` values ever get sent?

Agrair replied:
Actually, I'm a little confused here. There seems to be a few more `BinaryWriter` instances than there has to be, collectively in this method and the one super-method that calls it. Could someone please look at that? Since the `bool`s are now explicitly defined, I think all of them except the one passed in could just be removed.

Agrair commented on patches/tModLoader/Terraria/ModLoader/BiomeLoader.cs:
```diff
@@ -0,0 +1,142 @@
+
+					modBiome.Active = customReader.ReadBoolean();
+					//try { modPlayer.ReceiveCustomBiomes(reader) }
+				}
```
Again, this stuff with `BinaryWriter(MemoryStream)` is now going to cause overhead. I'll fix that in a bit.

Agrair commented on patches/tModLoader/Terraria/ModLoader/ModBiome.cs:
```diff
@@ -0,0 +1,68 @@
+		/// <summary>
+		/// Exactly the same as ModWorld.ModifyWorldGenTasks except it's tied to this ModBiome.
+		/// </summary>
+		public virtual void ModifyWorldGenTasks(List<GenPass> tasks, ref float totalWeight) {
```
Uhhh, I literally just copied this from `ModWorld`. I think it's getting phased out anyway, but I wasn't sure how to replace it. I was thinking of maybe autoloading nested types as `GenPass` objects?

Mirsario replied:
I don't think that biomes should deal with generation. Not in terraria.

Agrair replied:
So no generation hooks at all? Just let ModWorlds keep it?

kittenchilly replied:
Yeah, leave generation to ModWorld since vanilla doesn't really couple them with biomes in the first place

Agrair commented:
Whoops, I wasn't sure what that button did. I was hoping it meant "ask team member to review" 😆

Agrair commented on patches/tModLoader/Terraria/ModLoader/ModPlayer.cs:
```diff
@@ -24,6 +24,7 @@ public class ModPlayer:ModType
 		internal int index;
 
 		internal ModPlayer CreateFor(Player newPlayer) {
+			BiomeLoader.CopyBiomesTo(newPlayer);
```
Uhh, whoops.

Agrair commented:
Ok so I wrote this as if ModBiomes are responsible for data, but that’s absurd. Player obejcts should be passed in as parameters to the hooks and ModPlayer.modBiomes replaces with a simple Boolean array of biomes that registered as active.

Commit:
use a singleton design

Commit:
removed the world gen hook

Agrair commented on patches/tModLoader/Terraria/Player.TML.cs:
```diff
@@ -69,5 +69,18 @@ public partial class Player
+
+		/// <summary> Gets the instance of the specified ModBiome type. This will throw exceptions on failure. </summary>
+		/// <exception cref="IndexOutOfRangeException"/>
+		public bool IsModBiomeActive<T>() where T : ModBiome
```
I might change this to "InModBiome", just to make more sense to modders.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModBiome.cs:
```diff
@@ -0,0 +1,52 @@
+			BiomeLoader.Add(this);
+		}
+
+		internal void Update(Player player, ref bool value) {
```
I think it makes more sense to move this method into the loop of `BiomeLoader.Update`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModBiome.cs:
```diff
@@ -0,0 +1,52 @@
+		/// <summary>
+		/// Allows you to create special visual effects in the area around the player. For example, the blood moon's red filter on the screen or the slime rain's falling slime in the background. You must create classes that override Terraria.Graphics.Shaders.ScreenShaderData or Terraria.Graphics.Effects.CustomSky, add them in your mod's Load hook, then call Player.ManageSpecialBiomeVisuals. See the Exa ...
+		/// </summary>
+		public virtual void UpdateBiomeVisuals(Player player) {
```
multiple biomes may be simultaneously active. we should provide a priority mechanism to handle background drawing or music. 

Unlike player layers, here at may make sense to continually update the ordering, at least to find the 'most active' biome. I can imagine a system with several levels or categories, and then within categories a resolution via a weight hook (normally block count).

Eg, ice biome vs hallowed biome could be whichever block count is higher, but celestial pillar (event biome) takes priority

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/BiomeLoader.cs:
```diff
@@ -0,0 +1,92 @@
+
+		public static void SendCustomBiomes(Player player, BinaryWriter writer) {
+			for (int i = 0; i < player.modBiomeFlags.Length; i++) {
+				writer.Write(player.modBiomeFlags[i]);
```
bit packing please. Honestly the whole `modBiomeFlags` should probably be https://docs.microsoft.com/en-us/dotnet/api/system.collections.bitarray?view=netframework-4.5 instead of `bool[]`

Note that `BitArray` is optimised for performance. One day we might consider (or get yorai to consider) them for ID.Sets as well

Agrair replied:
Hmm, well for `BitArray`, the internal backing field isn't exposed at all (`m_array`). When I use `BitArray`, I just use reflection to retrieve it, but I don't think that's the right way to serialize it, is it?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/BiomeLoader.cs:
```diff
@@ -0,0 +1,92 @@
+		public static void CopyCustomBiomesTo(Player player, Player other) {
+			for (int i = 0; i < player.modBiomeFlags.Length; i++) {
+				other.modBiomeFlags[i] = player.modBiomeFlags[i];
+			}
```
`BitArray.Clone`?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/BiomeLoader.cs:
```diff
@@ -0,0 +1,92 @@
+				if (player.modBiomeFlags[i] ^ other.modBiomeFlags[i])
+					return false;
+			}
+			return true;
```
`BitArray.Equals`?

Agrair replied:
Tested it, and it doesn't look like `BitArray` implements a value based override for `Equals`, unfortunately.

Chicken-Bones replied:
huh, I checked the reference source and you're right. That's mega lame... Guess we can make our own?

Chicken-Bones reviewed (commented):
Good stuff

Commit:
Merge remote-tracking branch 'upstream/1.4' into 1.4_modbiome

Commit:
Merge branch '1.4' into 1.4_modbiome

Solxanich committed:
Address merge conflicts

Solxanich committed:
Merge pull request #5 from Agrair/tModLoader-1.4

T mod loader 1.4.2

Solxanich commented:
To Dos:
- [ ]  GlobalBiome neccessary? Move UpdateBiome there?
- [ ]  Complete :bool ModType.TryFind or something similar (returns true if player is inside the biome, false if not) - https://github.com/tModLoader/tModLoader/pull/1205#discussion_r524393010
- [ ]  Hook Bool BiomeConditions, return true if the player is in the biome, false if not (false default). Shows music, spawn conditions, etc
- [ ]  Hook OnInBiome, make stuff happen when a player is in the biome outside of background and music.
- [ ]  Set Background when in the biome
- [ ]  Set Map Background
- [ ]  Set Music (music should be able to change based on conditions, for stuff like night tracks, etc)
- [ ]  Set Localizable Name (that will be used anywhere where the biome name is shown)
- [ ]  Set Bestiary Tab icon, every biome in the game has one so ModBiome will automatically set it with a placeholder icon if not set here. (ModEvent or something similar should have this too). Should create a SpawnConditionBestiaryInfoElement as well.
- [ ]  Specify an ordering with respect to other biomes (including vanilla ones) similar to PlayerDrawLayers. A player can be in multiple biomes simultaneously, and they should all apply for drop rules and spawns, but things like background and music need to prioritise based on order.
- [ ] Convert modBiomeFlags to BitArray
- [ ] https://github.com/tModLoader/tModLoader/pull/1205#discussion_r525138467
- [ ] Check: https://github.com/tModLoader/tModLoader/pull/1205#discussion_r523705981
- [ ] Also fix CaptureBiome.cs - GetCaptureBiome -> #876
- [ ] Important vanilla methods to check: 
  - [ ] Player.GetPrimaryBiome. Extremely important, biome ids should be assigned automatically here and there should be a ModBiome.type or something that gets what this sets. See ordering above.
  - [ ] GameContent.Personalities.HelperInfo
  - [ ] GameContent.ShopHelper
  - [ ] GameContent.Bestiary.SpawnConditionBestiaryInfoElement (vanilla sets in BestiaryDatabaseNPCsPopulator, maybe add onto NPCSpawnHelper to do this?)

Solxanich committed:
A chunk of things for ModBiome

1) ModifyShopPrices hook for Personalities.HelperInfo
2) Add language grab for biomes in ShopHelper
3) Cleanup UpdateBiomes
4) bitArray part 1
5) Biome Atmosphere related stuff initializzed
6) Biome weight related stuff
7) Primary Biome Hijacked
8) Modifed IsModBiomeActive to be TryCheckIfInBiome(biome)

Solxanich committed:
Start work on setting up Backgrounds

There's a surprising amount of existing problems to fix in ModBackgrounds first.

Solxanich committed:
Implement IAtmosphericType and AtmosphericLoader

Refactors away duplicate code, and sets up for prioritized Client Atmosphere assignment based on a struct cached to the player

Solxanich committed:
Switch all ChooseStyle calls to use the cached, prioritized styles on player

That should complete prioritized mod backgrounds,
and both vanilla and mod prioritized music.

Should be ready for an initial peer review.

Solxanich committed:
Change from ModAtmosphereType to ModAVFX & Start ExampleBiome

Improvements recc from discord:
1) GetPrimaryBiome is overly complicated; look for improvements
2) Organization still needs improvement as a whole for tML, not sure will tackle under this as believe there is seperate issue for it.
3) Something something weights seem overly convoluted.

Commit:
Merge remote-tracking branch 'upstream/1.4' into 1.4_modbiome

Solxanich commented on patches/tModLoader/Terraria/ModLoader/AtmosphericLoader.cs:
```diff
@@ -0,0 +1,137 @@
+				shortList.Sort((s1, s2) => s1.weight.CompareTo(s2.weight));
+
+				foreach (var atmos in shortList) {
+					if (atmos.weight > weight) {
```
I need to remove this. It may cause issues and no benefit to having this check.
Thanks for pointing out Agrair

Agrair replied:
As a follow up, I think I fixed it, but it would be good if someone could [input on it](https://github.com/tModLoader/tModLoader/pull/1205/files#diff-9a1f12f29160d4c2b7c6a77b3ff9a11217e584089609fb35582ec0778bccfae3R113-R144)

Solxanich replied:
Fairly clear what it did, no worries. Looks good

Solxanich commented on patches/tModLoader/Terraria/ModLoader/AtmosphericLoader.cs:
```diff
@@ -0,0 +1,137 @@
+					if (atmos.weight > weight) {
+						weight = atmos.weight;
+
+						result.priority = atmos.type.Priority;
```
Agrair found bug:
biome: priority 1: music X1, background X2
event: priority 4: background X3.
Result: priority 4: background X3, music X1
Which means that either would have to wrap vanilla and deviate from how ModSystem did the ChooseMusic originally OR assume all derived classes have music. Notably, this only applies to music selection against vanilla today.
Given that the priority result is just cached, no reason we couldn't set it to match music at this point in time?

Agrair replied:
As a hotfix and possibly a permanent change, I did just [that](https://github.com/tModLoader/tModLoader/pull/1205/files#diff-9a1f12f29160d4c2b7c6a77b3ff9a11217e584089609fb35582ec0778bccfae3R140). However, we should leave this thread open as it is liable to change.

Commit:
Edited a couple of things (see desc)

AtmosphericLoader.cs -> AVFXLoader.cs
Player.currentModAtmosphere -> Player.currentAVFX
AVFXLoader.ModAtmosphere -> AVFXLoader.AVFXInstance
ModBiome.PostUpdateBiome -> ModBiome.BiomeVisuals
Removed some checks from ModPlayer relating to biome netcode
Cleaned up AVFXLoader.UpdateAVFX a little

Mirsario commented on ExampleMod/Content/Biomes/ExampleBiome.cs:
```diff
@@ -0,0 +1,51 @@
+
+		public override void ModifyShopPrices(HelperInfo helperInfo, ShopHelper shopHelperInstance) {
+			switch (helperInfo.npc.type) {
+				case 17:
```
NPCID please.

Mirsario commented on ExampleMod/Content/Biomes/ExampleBiome.cs:
```diff
@@ -0,0 +1,51 @@
+		}
+	}
+
+	public class ExampleUndergroundBiome : ModBiome
```
No reason for both to be in the same file. There's a separate folder for biomes after all.

Mirsario commented on patches/tModLoader/Terraria/ModLoader/ModAVFX.cs:
```diff
@@ -0,0 +1,46 @@
+		public virtual AVFXPriority Priority => AVFXPriority.None;
+
+		// Capture Camera property
+		public virtual CaptureBiome.TileColorStyle tileColorStyle => CaptureBiome.TileColorStyle.Normal;
```
What's with the lower case?

Commit:
Bestiary

I think. Example Mod may actually crash now as I was too lazy to make the backing textures (yet).

Solxanich commented on patches/tModLoader/Terraria/ModLoader/AVFXLoader.cs:
```diff
@@ -0,0 +1,161 @@
+			shortList.Sort(AtmosWeight.InvertedCompare);
+			int avfxFields = 0;
+
+			for (int i = 0; avfxFields < 7 && i < shortList.Count; i++) {
```
like the idea behind picking from the top in the shortList and exit early, but need to think about that surfaceBG and undergroundBG are in general mutually exclusive and may never be both set

Solxanich committed:
A couple of changes.

- swap defunct AVFXHelper for AVFXLoader
- Update BiomeLoader.GetPrimaryBiome to be less convoluted, and insert based on the AVFX priority within vanilla Player.GetPrimaryBiome.
TODO:
- Think about how priority works further for backgrounds; probably should also be functioning the same as music in terms of Priority.
- Look into the point Ozzatron raised on night corruption snow to figure out how ordering works with that sort of thing

Solxanich committed:
Add priority sorting on backgrounds to match Music

Solxanich committed:
Many updates to ExampleBiome

Implemented Backgrounds, needs more testing
Implemented Water Style, needs more testing

Verified CaptureCamera is fixed.
Just testing and verifying left. Almost there!

Solxanich committed:
Resolve conflicts with Main

Solxanich committed:
Merge pull request #7 from Agrair/tModLoader-1.4

T mod loader 1.4

Solxanich committed:
Minor updates & fixes

Remove ModifyShopPrices hook as frivolous
Not sure how far along bestiary is

Solxanich committed:
More Refactors to cleanup, and attempting to get Bestiary working

More testing needed

Solxanich committed:
Revert "More Refactors to cleanup, and attempting to get Bestiary working"

This reverts commit 6915692a2110a2798f46b75a9089e20cf466ae48.

Solxanich committed:
Removed weird unwanted patches & Recommit

Solxanich committed:
Now update main.cs when it isn't borked

Solxanich committed:
Minor Cleanup & bug fixes

Solxanich committed:
Tweaks on Prioritization to be smarter

Solxanich committed:
A bug fix, and notes on stuff that is broken

Solxanich committed:
Fix Backgrounds

- Fix autoloading from a subdirectory for backgrounds & sounds
- Fix Surface backgrounds hooks to be called again
- Update exampleMod to match

Solxanich committed:
Finish Bestiary & Thus the PR

Solxanich commented:
ModBiome PR is now tested and Complete.
- re-adds Mod Surface Backgrounds
- Fixes autoloading backgrounds from a subdirectory
- Fixes CaptureCamera for ModBiomes
- Adds and Cleans-up prioritization of backgrounds, water-styles to match music, including better class management of those loaders to accomplish this
- Adds bestiary support for Biomes
- Adds a ModBiome singleton to decrease god-classing in ModPlayer
- Ports ExampleBiomes to 1.4

Solxanich committed:
Merge pull request #8 from tModLoader/1.4

1.4 catchup

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModBiome.cs:
```diff
@@ -0,0 +1,90 @@
+		/// Return true if the player is in the biome.
+		/// </summary>
+		/// <returns></returns>
+		public virtual bool IsBiomeActive(Player player) => false;
```
should be distinguished from `IsActive` more easily. Perhaps this could be protected, and have an internal accessor for tML only? Something like `CalculateIsPlayerInBiome`?

Chicken-Bones commented on patches/tModLoader/Terraria/Player.TML.cs:
```diff
@@ -98,5 +99,29 @@ public partial class Player {
+		public bool InModBiome(ModBiome baseInstance)=> modBiomeFlags[baseInstance.ZeroIndexType];
+
+		public bool zonePurity = false;
+		public bool InZonePurity() {
```
needs a summary at least. Maybe it should be `NoBiomesActive`?

Chicken-Bones reviewed (commented):
Looks quite good!

Extra question: How are biomes used in NPC loot tables? Can we add an example?

Solxanich committed:
Commit Merge Result

Solxanich committed:
Merge pull request #10 from Agrair/tModLoader-1.4

T mod loader 1.4

Solxanich committed:
Merge pull request #9 from tModLoader/1.4

1.4 catchup

Solxanich committed:
Review comment Resolution Part1

Better AVFX docs
zonePurity docs

Solxanich committed:
Review Comment Resolution Part 2

- Greatly overhaul exampleMod documentation
- Cleanup a bit of residue BitArray tasks in BiomeLoader
- Remove bad Main.cs patch

Solxanich commented on ExampleMod/Content/Biomes/ExampleSurfaceBiome.cs:
```diff
@@ -0,0 +1,39 @@
+
+		public override bool IsBiomeActive(Player player) {
+			bool b1 = ModContent.GetInstance<ExampleBiomeTileCount>().exampleBlockCount >= 40;
+			bool b2 = Math.Abs(player.position.ToTileCoordinates().X - Main.maxTilesX / 2) < 40;
```
Changed to middle third of the world to give more room. 
Should give plenty of space; and in terms of if 40 was enough, it was roughly one screen length either direction of spawn, but doesn't hurt to make it bigger.

Solxanich commented on patches/tModLoader/Terraria/ModLoader/ModAVFX.cs:
```diff
@@ -0,0 +1,48 @@
+		/// <summary>
+		/// Returns the strength to which the AVFXEffects are active. 0% < val < 200%. Defaults to 100% (ie 100) 
+		/// </summary>
+		public virtual byte GetWeight(Player player) => 100;
```
I've added much more documentation to hopefully make it clear what it does.
Not sure why would want to return int, when GetCorrWeight is going to bound it anyways between 0 and 200, so any additional bytes are waste.

Chicken-Bones replied:
`int` is the preferred way to express 'this is just a number'. `byte` should only be used when the value is being stored or transmitted, in a place where the memory cost matters. It provides more to think about, is slightly harder to refactor, and makes general math harder (eg averaging). 

Similarly, it's accepted as best practice to use `int` in for loops or array indexers instead of `uint` even though the value cannot go below 0.

Solxanich commented:
> How are biomes used in NPC loot tables? Can we add an example

I have zero idea. If we want to look in to that, wouldn't that be better to do as a separate PR afterwards for an additional feature/investigation? As is, this contains several fixes for broken 1.3 features such as capture camera & surface backgrounds, so'd prefer not to defer for independently patchable enhancements.
But, can look in to if its a need under this PR.

Chicken-Bones commented:
> > How are biomes used in NPC loot tables? Can we add an example
> 
> I have zero idea, haven't changed anything afaik related to it. If we want to look in to that, wouldn't that be better to do as a separate PR afterwards for an additional feature/investigation? As is, this contains several fixes for broken 1.3 features such as capture camera & surface backgrounds and feels feature bloated as a PR, so'd prefer to defer for independently patchable enhancements.
> But, can look in to if its a need under this PR.

Happy for that to be a followup PR. I'm pretty happy with getting this merged to get modder feedback and testing.

Solxanich committed:
CorrWeight Update to bring in line with float percents & cleanup ExampleMod
